src/twisted/conch/scripts/ckeygen.py

```diff
@@ -202,10 +203,22 @@ def _defaultPrivateKeySubtype(keyType):
         return "PEM"
 
 
-def printFingerprint(options):
+def _getKeyOrDefault(options):
+    """
+    If C{options["filename"]} is None, prompt the user to enter a path
+    or attempt to set it to .ssh/id_rsa
+    """
```

Comment:
-----------

And as commented by JP, this function should also allow dependency injection, so that we don't need to path the input, but pass our own custom input method during the tests.
```suggestion
def _getKeyOrDefault(options, input_collector=None):
    """
    If C{options["filename"]} is None, prompt the user to enter a path
    or attempt to set it to .ssh/id_rsa
    """
    if input_collector is None:
        input_collector = input
```

-------

But also it should return the path, instead of mutating the input argument.

Reply:
The problem is that prompting the user before calling _getKeyOrDefault means we don't have a suggested/default path to include in the prompt. We can't just use a string, because that will only work for Windows or Linux, not both.

Reply:
I don't understand the issue.

Here is the complete implemenentation

```python
def _getKeyOrDefault(options, input_callback=input):
    """
    If C{options["filename"]} is None, prompt the user to enter a path
    or attempt to set it to .ssh/id_rsa.

    @param input_collector: A callback used to collect user input.
      Default to builting L{input} method.
    """
    filename = options["filename"]
    if not filename:
        filename = os.path.expanduser("~/.ssh/id_rsa")
        if platform.system() == "Windows":
            filename = os.path.expandvars(R"%HOMEPATH %\.ssh\id_rsa")
        filename = input_callback("Enter file in which the key is (%s): " % filename) or filename
    return 
```

Reply:
Updated.